struct_arm64.go

```diff
@@ -0,0 +1,202 @@
+
+package purego
+
+import (
+	"math"
+	"reflect"
+)
+
+func addStruct(v reflect.Value, numInts, numFloats, numStack *int, addInt, addFloat, addStack func(uintptr), keepAlive []interface{}) []interface{} {
+	if v.Type().Size() == 0 {
+		return keepAlive
+	}
+	// https://github.com/ARM-software/abi-aa/blob/main/sysvabi64/sysvabi64.rst
+	const (
+		NO_CLASS = 0b00
+		FLOAT    = 0b01
+		INT      = 0b11
+	)
+	if hva, hfa, size := isHVA(v.Type()), isHFA(v.Type()), v.Type().Size(); hva || hfa || size <= 16 {
+		// if this doesn't fit entirely in registers then
+		// each element goes onto the stack
+		if hfa && *numFloats+v.NumField() > numOfFloats {
+			*numFloats = numOfFloats
+		} else if hva && *numInts+v.NumField() > numOfIntegerRegisters() {
+			*numInts = numOfIntegerRegisters()
+		}
+
+		var val uint64
+		var shift byte
+		var flushed bool
+		class := NO_CLASS
+		var place func(v reflect.Value)
+		place = func(v reflect.Value) {
+			var numFields int
+			if v.Kind() == reflect.Struct {
+				numFields = v.Type().NumField()
+			} else {
+				numFields = v.Type().Len()
+			}
+			for k := 0; k < numFields; k++ {
+				var f reflect.Value
+				if v.Kind() == reflect.Struct {
+					f = v.Field(k)
+				} else {
+					f = v.Index(k)
+				}
+				if shift >= 64 {
+					shift = 0
+					flushed = true
+					if class == FLOAT {
+						addFloat(uintptr(val))
+					} else {
+						addInt(uintptr(val))
+					}
+				}
+				switch f.Type().Kind() {
+				case reflect.Struct:
+					place(f)
+				case reflect.Bool:
+					if f.Bool() {
+						val |= 1
+					}
+					shift += 8
+					class |= INT
+				case reflect.Uint8:
+					val |= f.Uint() << shift
+					shift += 8
+					class |= INT
+				case reflect.Uint16:
+					val |= f.Uint() << shift
+					shift += 16
+					class |= INT
+				case reflect.Uint32:
+					val |= f.Uint() << shift
+					shift += 32
+					class |= INT
+				case reflect.Uint64:
+					addInt(uintptr(f.Uint()))
+					shift = 0
+				case reflect.Int8:
+					val |= uint64(f.Int()&0xFF) << shift
+					shift += 8
+					class |= INT
+				case reflect.Int16:
+					val |= uint64(f.Int()&0xFFFF) << shift
+					shift += 16
+					class |= INT
+				case reflect.Int32:
+					val |= uint64(f.Int()&0xFFFF_FFFF) << shift
+					shift += 32
+					class |= INT
+				case reflect.Int64:
+					addInt(uintptr(f.Int()))
+					shift = 0
+				case reflect.Float32:
+					if class == FLOAT {
+						addFloat(uintptr(val))
+						val = 0
+						shift = 0
+					}
+					val |= uint64(math.Float32bits(float32(f.Float()))) << shift
+					shift += 32
+					class |= FLOAT
+				case reflect.Float64:
+					addFloat(uintptr(math.Float64bits(float64(f.Float()))))
+					shift = 0
+				case reflect.Array:
+					place(f)
+				default:
+					panic("purego: unsupported kind " + f.Kind().String())
+				}
+			}
+		}
+		place(v)
+		if !flushed {
+			if class == FLOAT {
+				addFloat(uintptr(val))
+			} else {
+				addInt(uintptr(val))
+			}
+		}
+	} else {
```

Comment:
This can also be split into the register version and the stack version for readability.